radix-engine/src/model/proof.rs

```diff
@@ -33,103 +45,314 @@ pub enum ProofError {
     ResourceContainerError(ResourceContainerError),
     /// Can't generate zero-amount or empty non-fungible set proofs.
     EmptyProofNotAllowed,
+    /// The base proofs are not enough to cover the requested amount or non-fungible ids.
+    InsufficientBaseProofs,
     /// Can't apply a non-fungible operation on fungible proofs.
     NonFungibleOperationNotAllowed,
+    /// Can't apply a fungible operation on non-fungible proofs.
+    FungibleOperationNotAllowed,
 }
 
 impl Proof {
-    // TODO: composite proofs
-    // TODO: proof auto drop
-    // TODO: thorough test partial/full/composite proofs
+    pub fn new_fungible(
+        resource_def_id: ResourceDefId,
+        restricted: bool,
+        total_amount: Decimal,
+        sources: HashMap<ProofSourceId, (Rc<RefCell<ResourceContainer>>, Decimal)>,
+    ) -> Result<Proof, ProofError> {
+        if total_amount.is_zero() {
+            return Err(ProofError::EmptyProofNotAllowed);
+        }
+
+        Ok(Self::Fungible {
+            resource_def_id,
+            restricted,
+            total_amount,
+            sources,
+        })
+    }
 
-    pub fn new(
+    pub fn new_non_fungible(
         resource_def_id: ResourceDefId,
-        resource_type: ResourceType,
         restricted: bool,
-        locked_total: AmountOrIds,
-        locked_details: Vec<(Rc<RefCell<ResourceContainer>>, AmountOrIds)>,
+        total_ids: BTreeSet<NonFungibleId>,
+        sources: HashMap<ProofSourceId, (Rc<RefCell<ResourceContainer>>, BTreeSet<NonFungibleId>)>,
     ) -> Result<Proof, ProofError> {
-        if match &locked_total {
-            AmountOrIds::Amount(amount) => amount.is_zero(),
-            AmountOrIds::Ids(ids) => ids.is_empty(),
-        } {
+        if total_ids.is_empty() {
             return Err(ProofError::EmptyProofNotAllowed);
         }
 
-        Ok(Self {
+        Ok(Self::NonFungible {
             resource_def_id,
-            resource_type,
             restricted,
-            locked_total,
-            locked_details,
+            total_ids,
+            sources,
         })
     }
 
+    pub fn compose(
+        proofs: &[Proof],
+        resource_def_id: ResourceDefId,
+        resource_type: ResourceType,
+    ) -> Result<Proof, ProofError> {
+        match resource_type {
+            ResourceType::Fungible { .. } => {
+                Self::compose_by_amount(proofs, None, resource_def_id, resource_type)
+            }
+            ResourceType::NonFungible => {
+                Self::compose_by_ids(proofs, None, resource_def_id, resource_type)
+            }
+        }
+    }
+
+    pub fn compose_by_amount(
+        proofs: &[Proof],
+        total_amount: Option<Decimal>,
+        resource_def_id: ResourceDefId,
+        resource_type: ResourceType,
+    ) -> Result<Proof, ProofError> {
+        // check resource type
+        if matches!(resource_type, ResourceType::NonFungible) {
```

Comment:
Is this check temporary? I feel like we still want to be able to compose amount proofs for non fungibles where the user doesn't care about the set of NonFungibleIds but only needs a specific amount.

Reply:
My original idea was not to support creating amount proofs from fungible resources, but realised that "amount" is just a way of expressing user intention, rather than fungibility. 

Will add support for this, the same way as in `take`  - check granularity and pick non-fungibles based on ID order.